### PR TITLE
Add transform flip

### LINKED ISSUE #116702: Add a way to flip/mirror an widget
## Use case

We need a named option to flip widgets horizontally or vertically. It's been mentioned in certain discussions that flutter does not provide a way to flip/mirror widgets. Though we can mirror a child by applying negative scaling, but it would be more appealing to have a dedicated option to flip/mirror widgets.

## Proposal

Add a named constructor in the Transform widget to flip/mirror it's child

### PR DESCRIPTION
**Pull request by saminarp**
Adds Transform.flip to allow flipping/mirroring child widgets

Fixes https://github.com/flutter/flutter/issues/116702

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by cedvdb** [comment]
Is this a 3D lookalike flip ? If that's the case shouldn't skew a little ?

**Comment by kevmoo** [comment]
Great first PR, @saminarp!! 👏 👏 👏

File: packages/flutter/lib/src/widgets/basic.dart
Comment (justinmc) [inline packages/flutter/lib/src/widgets/basic.dart:1455]: I think my gut says this is ok as-is, but I think we should discuss: What if we added an assert enforcing that flipX and flipY can't both be false?

On one hand it's somewhat confusing to have a widget called `Transform.flip` that doesn't flip anything.  Users may try it without knowing about flipX/flipY and be confused that nothing flips.

On the other hand, it may be convenient to be able to use a Transform.flip that flips based on some state without having to exclude it from the widget tree.

What do you think?

File: packages/flutter/lib/src/widgets/basic.dart
Comment (justinmc) [inline packages/flutter/lib/src/widgets/basic.dart:1449]: Another thing to think about: Is there a better way to represent this?  For example, an enum with values like x, y, both, and none.  Kind of similar to [Axis](https://master-api.flutter.dev/flutter/painting/Axis.html).

I'm not saying that flipX/flipY aren't the best way to do this already, just starting a discussion again.

File: packages/flutter/test/widgets/transform_test.dart
Comment (justinmc) [inline packages/flutter/test/widgets/transform_test.dart:834]: "true" => "isTrue"

File: packages/flutter/lib/src/widgets/basic.dart
Comment (justinmc) [inline packages/flutter/lib/src/widgets/basic.dart:1455]: What effect does the -1 z value have?

File: packages/flutter/lib/src/widgets/basic.dart
Comment (justinmc) [inline packages/flutter/lib/src/widgets/basic.dart:1427]: Nit: "Creates a widget that flips/mirrors its child in place along the 2D plane."

I'm thinking that we should probably mention somewhere in these docs that the flip happens "in place", so the widget doesn't get moved.  Or maybe if you don't want to change this first sentence you could mention elsewhere that it defaults to Alignment.center or something like that.  Whatever you think is best.

File: packages/flutter/test/widgets/transform_test.dart
Comment (justinmc) [inline packages/flutter/test/widgets/transform_test.dart:790]: Nit: I would usually turn this test into a `group` and then turn each `pumpWidget` below into a separate test.  That way you give each feature tested a chance to fail during one run and give a bit more info.  Up to you if you think it's useful or not here, though.

File: packages/flutter/test/widgets/transform_test.dart
Comment (justinmc) [inline packages/flutter/test/widgets/transform_test.dart:879]: Nit: One small thing that might make this test somewhat fragile is that we're always expecting tappedRed to be true, so this test would pass for a big gesture detector that didn't flip at all.  It would be more reassuring if you could test that the tap did not cause tappedRed to become true when nothing is flipped.  Maybe do something like test that tapping topRight, bottomLeft, and bottomRight don't set tappedRed to true when there is no flip?

Up to you though, maybe it's not worth it to add that complexity.

**Review by justinmc — COMMENTED** [review]
Just some nits and questions/discussions here before I approve.

I think I'm on board with adding this constructor, it makes sense to me to have a simple way to do this.

File: packages/flutter/lib/src/widgets/basic.dart
Comment (saminarp) [reply packages/flutter/lib/src/widgets/basic.dart:1455]: We could use a List<Axis> and fill it with the axes needed, empty list would mean no flip, but there would need to be a list and we could assert that

Comment (saminarp) [reply packages/flutter/lib/src/widgets/basic.dart:1455]: I applied matrix rotation and got this. So I thought I should preserve the value. It does not however affect the appearance as far as I noticed
CleanShot 2022-12-10 at 03 30 34@2x

File: packages/flutter/test/widgets/transform_test.dart
Comment (saminarp) [reply packages/flutter/test/widgets/transform_test.dart]: I have just done this in my recent commit

File: packages/flutter/test/widgets/transform_test.dart
Comment (saminarp) [reply packages/flutter/test/widgets/transform_test.dart]: I am a little behind schedule in a few of my college assignments. So, in my opinion, it is not worth adding the complexity at this time. 

Let me know your thoughts :)

**Comment by saminarp** [comment]
> Great first PR, @saminarp!! 👏 👏 👏

Thank you so much!

**Review by justinmc — APPROVED** [review]
LGTM 👍  assuming that the tests pass.

I restarted a few that were canceled, but it seems like you'll need to push a new commit to get the gold tests to re-run.  Can you push a merge commit?

CC @gspencergoog for secondary review

File: packages/flutter/lib/src/widgets/basic.dart
Comment (gspencergoog) [reply packages/flutter/lib/src/widgets/basic.dart:1458]: You shouldn't need to supply -1 in Z, you're just doing a 2D flip. I'd just always set it to 1.

File: packages/flutter/lib/src/widgets/basic.dart
Comment (gspencergoog) [reply packages/flutter/lib/src/widgets/basic.dart:1452]: I think two bools here is probably fine, given that it's a specialized constructor already.  I can't think of any other ways that "flip" might be interpreted here, unless we also want to incorporate 90 degree rotations (which I don't think we do), and making it an enum makes it harder to write a call to the constructor that derives from a boolean (which is probably the main use case here).

File: packages/flutter/lib/src/widgets/basic.dart
Comment (gspencergoog) [inline packages/flutter/lib/src/widgets/basic.dart:1433]: You might mention that flipping it in both is equivalent to a 180 degree rotation.

File: packages/flutter/lib/src/widgets/basic.dart
Comment (gspencergoog) [inline packages/flutter/lib/src/widgets/basic.dart:1436]: What happens if the alignment is `Alignment.upperLeft` and both are flipped?  Does it just draw nothing?  Or does it draw outside of its bounding box?

Comment (gspencergoog) [reply packages/flutter/lib/src/widgets/basic.dart:1436]: Actually, the more I think about it, isn't any alignment other than `center` going to draw nothing, or draw outside of its bounding box as long as one flip is true? That seems undesirable, since both are surprising.

Comment (nayeemtby) [reply packages/flutter/lib/src/widgets/basic.dart:1436]: It draws outside the bounding box if alignment is something other than `center`. Being restrictive, we could just fix the alignment to `Alignment.center` and not provide the option to change it at all. But allowing it may help someone desiring the undesirable results. But they could also do it using the main `Transform()` constructor. But again, the purpose of this `flip` constructor is to just simplify and be descriptive of what is being done. What do you prefer?

File: packages/flutter/test/widgets/transform_test.dart
Comment (nayeemtby) [reply packages/flutter/test/widgets/transform_test.dart:791]: Let's keep it as-is.

File: packages/flutter/lib/src/widgets/basic.dart
Comment (nayeemtby) [reply packages/flutter/lib/src/widgets/basic.dart]: Making Z constant to reduce unnecessary computation 👍

File: packages/flutter/lib/src/widgets/basic.dart
Comment (gspencergoog) [reply packages/flutter/lib/src/widgets/basic.dart:1436]: In that case, I think I'd prefer if we just eliminate the `alignment` argument altogether and just fix it at Alignment.center. It's surprising to have it draw outside of its bounding box, and if they want to do that, they can do it using the main constructor. We can always add it later if a compelling use case arises.

File: packages/flutter/lib/src/widgets/basic.dart
Comment (gspencergoog) [reply packages/flutter/lib/src/widgets/basic.dart:1427]:
```suggestion
  /// Creates a widget that mirrors its child about the widget's center point.
```

I actually don't think you need to specify that it's a 2D plane: everything in Flutter is assumed to be in 2D. It would be remarkable if it were a 3D transform, which is obviously totally possible with a 4x4 matrix, but all of our other transforms are 2D transforms, so it will be assumed that this one is too.

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

Looks good, thanks for putting up with all of our comments. :-)

**Comment by nayeemtby** [comment]
> Looks good, thanks for putting up with all of our comments. :-)

Standard operating procedure 🧐

**Comment by nayeemtby** [comment]
@gspencergoog Can you add the label again? It was removed and now all is green.

**Pull request merged**
